s3parq/publish_redshift.py

```diff
@@ -27,6 +27,103 @@ def _redshift_name_validator(*args):
+    '''
+    filepath = filename.split('/')
+    final_partitions = [_dir for _dir in filepath if '=' in _dir]
+    return final_partitions
+
+
+def _format_partition_strings_for_sql(partitions: [str]) -> [str]:
+    '''
+    Formats a list of S3 partition strings for SQL statements
+    Args:
+        partitions ([str]): list of strings representing the partitions in S3
+    ----
+    Returns:
+        formatted_partitions ([str]): list of the same partitions with quotes
+        to use for SQL
+    --------
+    Example:
+        Args:
+            partitions = ['hamburger=abcd', 'hot_dog=1234']
+        ----
+        Returns:
+            formatted_partitions = ["hamburger='abcd'", "hot_dog='1234'"]
+    '''
+    formatted_partitions = []
+    for p in partitions:
+        key, value = p.split('=')
+        value = f"'{value}'"
+        formatted_partitions.append(f'{key}={value}')
+    return formatted_partitions
+
+
+def _last_index_containing_substring(the_list: [str], substring: str) -> int:
+    '''
+    Returns index of last string that contains a substring within a list.  If there's no match, it returns
+    the length of the list + 1 because we want the function that uses it "_get_partition_location"
+    to not identify any partitions if none exists
+    Args:
+        the_list ([str]): list of strings, probably representing the partitions
+        substring (str): string to look for in the_list
+    ----
+    Returns:
+        i (int): The index in the list that contains the first instance of the substring
+        returns -1 if no matches
+    --------
+    Example:
+        Args: 
+            the_list = ['path', 'to', 'data', 'hamburger=abcd', 'hot_dog=1234', 'the_file.parquet]
+            substring = '='
+        ----
+        Returns:
+            i = 3
+        
+    '''
+    for s in reversed(the_list):
+        if substring in s:
+            return len(the_list) - the_list.index(s)
+    return len(the_list) + 1
+
+
+def _get_partition_location(filepath: str):
+    '''
+    Gets location of data in S3 for partitioning in S3.  You need to know the path to 
+    the first partition in order to make a proper Spectrum partition w/ SQL-Redshift
+    Args:
+        filepath (str): path to a file in S3 that is partitioned
+    Returns:
+        final_path (str): path within S3 bucket that has the last partition of a filepath
+    Example:
+        Args: 
+            filepath = 'path/to/data/hamburger=abcd/hot_dog=1234/abcd1234.parquet'
+        Returns:
+            final_path = 'path/to/data/hamburger=abcd/hot_dog=1234'
+    '''
+    separate_dirs = filepath.split('/')
+    last_partition = _last_index_containing_substring(separate_dirs, "=")
+    final_set = separate_dirs[:-last_partition + 1] #  I think this is more confusing than it has to be
+    final_path = '/'.join(final_set)
+    if final_path == '':
+        raise ValueError(f'No partitions in this filepath {filepath}')
```

Comment:
Triple checking but if there are no partitions sent this is never hit right? Didn't see a check or try except but also I'm blind